azurerm/internal/services/datamigration/resource_arm_data_migration_service.go

```diff
@@ -0,0 +1,247 @@
+	}
+	if t, ok := d.GetOk("tags"); ok {
+		parameters.Tags = tags.Expand(t.(map[string]interface{}))
+	}
+
+	future, err := client.CreateOrUpdate(ctx, parameters, resourceGroup, name)
+	if err != nil {
+		return fmt.Errorf("Error creating Data Migration Service (Service Name %q / Group Name %q): %+v", name, resourceGroup, err)
+	}
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for creation of Data Migration Service (Service Name %q / Group Name %q): %+v", name, resourceGroup, err)
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Data Migration Service (Service Name %q / Group Name %q): %+v", name, resourceGroup, err)
+	}
+	if resp.ID == nil {
+		return fmt.Errorf("Cannot read Data Migration Service (Service Name %q / Group Name %q) ID", name, resourceGroup)
+	}
+	d.SetId(*resp.ID)
+
+	return resourceArmDataMigrationServiceRead(d, meta)
+}
+
+func resourceArmDataMigrationServiceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).DataMigration.ServicesClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	name := id.Path["services"]
+
+	resp, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] Data Migration Service %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading Data Migration Service (Service Name %q / Group Name %q): %+v", name, resourceGroup, err)
+	}
+
+	d.Set("id", resp.ID)
+	if location := resp.Location; location != nil {
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+	d.Set("resource_group_name", resourceGroup)
+	d.Set("name", name)
+	if serviceProperties := resp.ServiceProperties; serviceProperties != nil {
+		d.Set("provisioning_state", string(serviceProperties.ProvisioningState))
+		d.Set("virtual_subnet_id", serviceProperties.VirtualSubnetID)
+	}
+	if err := d.Set("sku_name", resp.Sku.Name); err != nil {
+		return fmt.Errorf("Error setting `sku_name`: %+v", err)
+	}
+	d.Set("type", resp.Type)
+	d.Set("kind", resp.Kind)
+
+	return tags.FlattenAndSet(d, resp.Tags)
+}
+
+func resourceArmDataMigrationServiceUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).DataMigration.ServicesClient
+	ctx, cancel := timeouts.ForUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	name := id.Path["services"]
+
+	parameters := datamigration.Service{}
+
+	if t, ok := d.GetOk("tags"); ok {
+		parameters.Tags = tags.Expand(t.(map[string]interface{}))
```

Comment:
This should be `Tags: tags.Expand(t),` ?

Reply:
Because `d.GetOk()` returns an `interface{}` while `tags.Expand()` expect a `map[string]interface{}`. Need a type cast here to make compiler happy.

Reply:
Given ever where else we don't check if this exists or not i think this can be:
```
parameters := datamigration.Service{
	Tags := tags.Expand(d.Get("tags").(map[string]interface{}))
```